### PR TITLE
Moved open_session call from RequestContext.push to wsgi_app

### LINKED ISSUE #1528: Application/request stack corruption if unhandled exception is raised in session interface
RequestContext push is called in [wsgi_app method](https://github.com/mitsuhiko/flask/blob/master/flask/app.py#L1952), outside try..finally block. In [push() method](https://github.com/mitsuhiko/flask/blob/master/flask/ctx.py#L332) application's open_session call takes place, which can possibly lead to exception. But, since original call is not in try..finally block, removal of stack top never takes place. This leads to following problems:
### 1. Application context stack never updates again

RequestContext.push() only [checks](https://github.com/mitsuhiko/flask/blob/master/flask/ctx.py#L316) that application context exists or belongs to current application, but same rules apply to previous context.
### 2. Request context stack grows

Again, request context stack is pushed, but never removed. Actually just a memory leak :)
### 3. Misleading code comment about open_session usage

Inspired by [this comment](https://github.com/mitsuhiko/flask/blob/master/flask/ctx.py#L328), i started to develop multi-tenant application, and found nothing but disappointment. **You can't actually access request_context before open_session interface call.** This means, you can't configure which database to use before opening a session.
## My proposition.

Move open_session call to full_dispatch_request, and add interfaces and methods, similar to before_request - before_open_session, or call before_request before opening session (may break code that depends on session interface)

### PR DESCRIPTION
**Pull request by WeirdCarrotMonster**
Fixes #1528

New approach has only one drawback: in any custom tests that require sessions, you need to call `ctx_open_session` directly. I fixed existing tests in flask, but any user test may fail.

### CODE REVIEW

**Comment by keyan** [comment]
@WeirdCarrotMonster I'm concerned about breaking backwards compatibility. Is there anyway to avoid this? Did you update the docs to reflect your proposed change?

**Comment by WeirdCarrotMonster** [comment]
It's not as compatible as i'd like it to be. The problem is: current approach on managing RequestContext and sessions is wrong. I'm not sure if it can be fixed without breaking compatibility for some people, especially if they use `test_request_context` and `session` in their code.

I haven't updated the docs yet, since i hope there is better solution.

**Comment by voxmens** [comment]
would it be possible to reimplement those functions using the newer apis?

**Comment by ThiefMaster** [comment]
Also, please clean up the PR a bit. There shouldn't be merge commits in a PR at all (you always rebase feature branches to sync them with upstream) and the various test-related commits could be squashed into a single one.

File: tests/test_appctx.py [hunk 1]
```diff
@@ -43,6 +43,34 @@ def test_request_context_means_app_context():
...
+    app = flask.Flask(__name__)
+
+    class BadSessionInterface(object):
+
+        """Fails once."""
+        fail = True
+
+        def open_session(self, *args, **kwargs):
+            if self.fail:
+                setattr(flask.g, "test_g_attr", 1)
```
Comment (ThiefMaster) [inline tests/test_appctx.py]: Why setattr and not simply `flask.g.test_g_attr = 1`?

File: tests/test_appctx.py [hunk 2]
```diff
@@ -43,6 +43,34 @@ def test_request_context_means_app_context():
...
+                raise Exception()
+        def make_null_session(self, *args, **kwargs):
+            return None
+
+    app.session_interface = BadSessionInterface()
+
+    try:
+        with app.test_request_context():
+            app.ctx_open_session()
+    except:
```
Comment (ThiefMaster) [inline tests/test_appctx.py]: even in a test case, `except: pass` is a **terrible** idea.

File: tests/test_appctx.py [hunk 3]
```diff
@@ -43,6 +43,34 @@ def test_request_context_means_app_context():
...
+        pass
+
+    with app.test_request_context():
+        app.ctx_open_session()
+        assert getattr(flask.g, "test_g_attr", None) is None
```
Comment (ThiefMaster) [inline tests/test_appctx.py]: `flaks.g.get('test_g_attr')`

File: tests/test_appctx.py (hunk 2 above)
Comment (untitaker) [inline tests/test_appctx.py]: Use pytest.raises

On 3 August 2015 12:18:47 CEST, Adrian <notifications@github.com> wrote:
>> +        fail = True
>> +
>> +        def open_session(self, *args, **kwargs):
>> +            if self.fail:
>> +                setattr(flask.g, "test_g_attr", 1)
>> +                self.fail = False
>> +                raise Exception()
>> +        def make_null_session(self, *args, **kwargs):
>> +            return None
>> +
>> +    app.session_interface = BadSessionInterface()
>> +
>> +    try:
>> +        with app.test_request_context():
>> +            app.ctx_open_session()
>> +    except:
>
>even in a test case, `except: pass` is a **terrible** idea.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/mitsuhiko/flask/pull/1538/files#r36072407

-- 
Sent from my phone. Please excuse my brevity.

**Pull request closed without merging**